src/main/webapp/app/editor/ace/editor-ace.component.ts

```diff
@@ -0,0 +1,199 @@
+        const sessionKeys = Object.keys(this.editorFileSessions);
+        const unsavedFiles = sessionKeys.filter(session =>
+            this.editorFileSessions[session].unsavedChanges === true).length;
+
+        if (unsavedFiles > 0) {
+            if (this.onSaveStatusChange) {
+                this.onSaveStatusChange({
+                    isSaved: false,
+                    saveStatusLabel: '<i class="fa fa-circle-o-notch fa-spin text-info"></i> <span class="text-info">Unsaved changes in ' + unsavedFiles + ' files.</span>'
+                });
+            }
+        } else {
+            if (this.onSaveStatusChange) {
+                this.onSaveStatusChange({
+                    isSaved: true,
+                    saveStatusLabel: '<i class="fa fa-check-circle text-success"></i> <span class="text-success"> All changes saved.</span>'
+                });
+            }
+        }
+    }
+
+    /**
+     * Fetches the requested file by filename and opens a new editor session for it (if not yet done)
+     * @param fileName: Name of the file to be opened in the editor
+     */
+    loadFile(fileName: string) {
+        /** Query the repositoryFileService for the specified file in the repository */
+        this.repositoryFileService.get(this.participation.id, fileName).subscribe(fileObj => {
+
+            if (!this.editorFileSessions[fileName]) {
+                this.editorFileSessions[fileName] = {};
+                this.editorFileSessions[fileName].code = fileObj.fileContent;
+            }
+            /**
+             * Assign the obtained file content to the editor and set the ace mode
+             * Additionally, we resize the editor window and set focus to it
+             */
+            this.editorText = fileObj.fileContent;
+            // This fetches a list of all supported editor modes and matches it afterwards against the file extension
+            const aceModeList = ace.require('ace/ext/modelist');
+            const aceMode = aceModeList.getModeForPath(fileName);
+            this.editor.setMode(aceMode);
+            this.editor.getEditor().resize();
+            this.editor._editor.focus();
+        }, err => {
+            console.log('There was an error while getting file', this.fileName, err);
+        });
+    }
+
+    /**
+     * @function saveFile
+     * @desc Saved the currently selected file; is being called when the file is changed (onFileChanged)
+     * @param fileName: name of currently selected file
+     */
+    saveFile(fileName: string) {
+        if (this.onSaveStatusChange) {
+            this.onSaveStatusChange({
+                isSaved: false,
+                saveStatusLabel: ' <i class="fa fa-circle-o-notch fa-spin text-info"></i><span class="text-info"> Saving file.</span>'
+            });
+        }
+
+        this.repositoryFileService.update(this.participation.id,
+            fileName,
+            this.editorFileSessions[fileName].code)
+            .debounceTime(3000)
```

Comment:
please store 3000 as constant in the beginning of the file and give it a descriptive name